lighthouse-core/audits/seo/structured-data-automatic.js

```diff
@@ -0,0 +1,163 @@
+  /** Explanatory message stating what percentage of JSON-LD structured data snippets are invalid */
+  displayValue: '{validSnippetProportion, number, percent} valid snippets',
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+class StructuredDataAutomatic extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'structured-data-automatic',
+      title: str_(UIStrings.title),
+      failureTitle: str_(UIStrings.failureTitle),
+      description: str_(UIStrings.description),
+      requiredArtifacts: ['ScriptElements'],
+    };
+  }
+
+  /**
+   * @param {LH.Artifacts} artifacts
+   * @return {Promise<LH.Audit.Product>}
+   */
+  static async audit(artifacts) {
+    const jsonLDElements = artifacts.ScriptElements.filter(
+      script => script.type === 'application/ld+json' && !!script.content);
+
+    if (jsonLDElements.length === 0) {
+      return {
+        notApplicable: true,
+        score: 1,
+      };
+    }
+
+    const validatedSnippets = await Promise.all(jsonLDElements.map(async (element) => {
+      // We don't want to show empty lines around the snippet
+      const content = (element.content || '').trim();
+
+      return {
+        devtoolsNodePath: element.devtoolsNodePath,
+        content,
+        errors: await validateJsonLD(content),
+      };
+    }));
+    // Show invalid snippets at the top
+    validatedSnippets.sort((a, b) => {
+      return b.errors.length - a.errors.length;
+    });
+
+    const renderedSnippets = await Promise.all(
+      validatedSnippets.map(snippetWithErrors => renderValidatedSnippet(snippetWithErrors))
+    );
+    const details = Audit.makeListDetails(renderedSnippets);
+
+    const invalidSnippets = validatedSnippets.filter(vs => vs.errors.length > 0);
+    const validSnippets = validatedSnippets.filter(vs => vs.errors.length === 0);
+    const displayValue = str_(UIStrings.displayValue, {
+      validSnippetProportion: validSnippets.length / jsonLDElements.length,
+    });
+
+    return {
+      score: invalidSnippets.length === 0 ? 1 : 0,
+      details,
+      displayValue,
+    };
+  }
+}
+
+/**
+ * @param {{content: string, devtoolsNodePath: string, errors: LH.StructuredData.ValidationError[]}} validatedSnippet
+ */
+function renderValidatedSnippet(validatedSnippet) {
+  const {content, devtoolsNodePath, errors} = validatedSnippet;
+
+  let parsedContent;
+  let topLevelType;
+  let topLevelName;
+  try {
+    parsedContent = JSON.parse(content);
+    topLevelType = parsedContent['@type'];
+    topLevelName = parsedContent['name'];
+  } catch (err) {}
+
+  let title = '';
+  if (topLevelName && topLevelType) {
+    title = `${topLevelType}: ${topLevelName}`;
+  } else if (topLevelType) {
+    title = `@type ${topLevelType}`;
+  } else {
+    title = 'Invalid JSON-LD element';
+  }
+  title += ` (${errors.length} Error${errors.length !== 1 ? 's' : ''})`;
```

Comment:
can we i18n this sucker?

Reply:
Yeah maybe. I'm not quite sure how to do it, is there some way to put the `snippetName ? snippetName: 'Invalid JSON-LD element'` in the i18n string? It seems like we can't concatenate individual i18n strings because the final conversion is done on the full strings.

I wonder if we even need to, given that the actual error messages are in English and if the snippet can be parsed it'll use the English schema.org name.

Reply:
ah yeah the mingling with lib error messages messes with everything huh, nevermind

the manual pluralization might stick out to someone in the future, mind throwing a quick comment? :)

Reply:
Added one 🙂